colpali_engine/trainer/contrastive_trainer.py

```diff
@@ -1,12 +1,180 @@
+from typing import Iterator, List, Optional
+
+import numpy as np
 import torch
-from transformers import Trainer
+from datasets import DatasetDict
+from torch.utils.data import BatchSampler, ConcatDataset, DataLoader, Dataset
+from transformers import Trainer, is_datasets_available
+from transformers.trainer_utils import seed_worker
+
+
+class SingleDatasetBatchSampler(BatchSampler):
```

Comment:
I think we should implement some tests for this Sampler if possible. It has a quite complex logic.